### PR TITLE
fix: ignore useSyncExternalStore updates during unsubscribe

### PR DESCRIPTION
**Pull request by srpatcha**
## Summary

Prevent a `useSyncExternalStore` subscription callback from checking for updates after its subscription has begun cleanup. The cleanup marks the subscription inactive before calling the store's `unsubscribe` function, so a listener invoked synchronously during unsubscribe cannot read from a stale store instance.

Adds a focused regression test covering an unsubscribe function that invokes its captured listener.

## Test plan

- `yarn test useSyncExternalStore-test --runInBand` (5 passed)
- `yarn test-stable useSyncExternalStore-test --runInBand` (5 passed)
- `node scripts/tasks/eslint.js packages/react-reconciler/src/ReactFiberHooks.js packages/react-reconciler/src/__tests__/useSyncExternalStore-test.js`
- `node scripts/prettier/index.js check-changed`
- `git diff --check`

### CODE REVIEW

**Comment by srpatcha** [comment]
Cleaned up + rebased onto current main (commit `171ed8c`):
- ✅ Rebased onto current `main` (was BEHIND with merge conflict)
- ✅ Dropped the empty-content `a8dd8395` commit (its actual typeName code was lost in an earlier squash, leaving only file-mode changes that conflicted with upstream's release-script deletions)
- ✅ Re-scoped PR to the 3 substantive commits: `useThrottledCallback` hook + `useSyncExternalStore` stale-closure fix + Jest tests
- ✅ Updated title and body to reflect actual scope

Ready for re-review!

**Comment by eps1lon** [comment]
Can you add a test for the stale closure issue in useSyncExternalStore following our current testing practice? The `useThrottledCallback ` addition can be removed from this PR. This implementation belongs in userland.

**Comment by srpatcha** [comment]
Addressed the feedback in `12439646e9bb55261c214801556d8202bedb02ed`: removed `useThrottledCallback` from the PR and added a focused `useSyncExternalStore` regression test for a listener invoked synchronously during unsubscribe.

Exact local results:
- `yarn test useSyncExternalStore-test --runInBand`: 1 suite passed, 5 tests passed, 0 snapshots
- `yarn test-stable useSyncExternalStore-test --runInBand`: 1 suite passed, 5 tests passed, 0 snapshots
- Targeted ESLint on both changed files: passed
- `node scripts/prettier/index.js check-changed`: passed
- `git diff --check`: passed

I also verified the regression fails without the guard: `getSnapshot` is called once during unsubscribe.

File: packages/react-reconciler/src/__tests__/useSyncExternalStore-test.js [hunk 1]
```diff
@@ -141,6 +141,31 @@ describe('useSyncExternalStore', () => {
...
   }
 
+  it('does not check for updates when unsubscribe calls the listener', async () => {
+    let currentValue = 'A';
+    const getSnapshot = jest.fn(() => currentValue);
+    const subscribe = listener => {
+      return () => {
+        currentValue = 'B';
+        listener();
+      };
```
Comment (eps1lon) [inline packages/react-reconciler/src/__tests__/useSyncExternalStore-test.js:151]: What's the real-world scenario where you would update the store during unsubscribe **and** call the listener that should've been unsubscribed during the  unsubscribe call?

Comment (srpatcha) [reply packages/react-reconciler/src/__tests__/useSyncExternalStore-test.js:151]: I do not have a concrete production store that does both. The regression test models re-entrant teardown: `unsubscribe` mutates the source and synchronously invokes the captured callback before returning. In the common implementations I checked, unsubscribe only detaches, so this is a defensive edge case rather than a demonstrated real-world integration. If React does not want to support callbacks during teardown, then the guard is not justified by this test alone.

**Review by eps1lon — CHANGES_REQUESTED** [review]
Waiting for a legit scenario (see https://github.com/react/react/pull/36343#discussion_r3953869019). So far this looks like incorrect usage to begin with
